src/vm/amsi.cpp

```diff
@@ -0,0 +1,147 @@
+        {
+            // KB2533623 introduced the LOAD_LIBRARY_SEARCH_SYSTEM32 flag. It also introduced
+            // the AddDllDirectory function. We test for presence of AddDllDirectory as an
+            // indirect evidence for the support of LOAD_LIBRARY_SEARCH_SYSTEM32 flag.
+            HMODULE kernel32 = GetModuleHandleW(WINDOWS_KERNEL32_DLLNAME_W);
+            if (kernel32 != nullptr)
+            {
+                loadlibrarySearchSystem32Enabled = 0;
+                if (GetProcAddress(kernel32, "AddDllDirectory") != nullptr)
+                {
+                    // This system supports the LOAD_LIBRARY_SEARCH_SYSTEM32 flag
+                    loadlibrarySearchSystem32Enabled = 1;
+                }
+            }
+        }
+
+        return (loadlibrarySearchSystem32Enabled == 1);
+    }
+}
+
+// Here we will invoke into AmsiScanBuffer, a centralized area for non-OS
+// programs to report into Defender (and potentially other anti-malware tools).
+// This should only run on in memory loads, Assembly.Load(byte[]) for example.
+// Loads from disk are already instrumented by Defender, so calling AmsiScanBuffer
+// wouldn't do anything.
+bool Amsi::IsBlockedByAmsiScan(PVOID flatImageBytes, COUNT_T size)
+{
+    STANDARD_VM_CONTRACT;
+
+    if (!InitializeLock())
+        return false;
+
+    // Lazily initialize AMSI because it is very expensive
+    {
+        CRITSEC_Holder csh(s_csAmsi);
+
+        // Cache that we failed if this didn't work so we don't keep trying to reinitialize
+        static bool amsiInitializationAttempted = false;
+        if (s_amsiContext == nullptr && !amsiInitializationAttempted)
+        {
+            // Only OSes that support LOAD_LIBRARY_SEARCH_SYSTEM32 have AMSI
+            if (HasLoadLibrarySearchSystem32())
+            {
+                HMODULE amsi = CLRLoadLibraryEx(W("amsi.dll"), nullptr, LOAD_LIBRARY_SEARCH_SYSTEM32);
+                if (amsi != nullptr)
+                {
+                    PAMSI_AMSIINITIALIZE_API AmsiInitialize = (PAMSI_AMSIINITIALIZE_API)GetProcAddress(amsi, "AmsiInitialize");
+                    if (AmsiInitialize != nullptr)
+                    {
+                        HAMSICONTEXT amsiContext = nullptr;
+                        if (AmsiInitialize(W("coreclr"), &amsiContext) == S_OK)
+                        {
+                            AmsiScanBuffer = (PAMSI_AMSISCANBUFFER_API)GetProcAddress(amsi, "AmsiScanBuffer");
+                            if (AmsiScanBuffer != nullptr)
+                            {
+                                s_amsiContext = amsiContext;
+                            }
+                        }
+                    }
+                }
+            }
+
+            amsiInitializationAttempted = true;
+        }
+    }
+
+    if (s_amsiContext == nullptr || AmsiScanBuffer == nullptr)
+        return false;
+
+    DWORD result;
+    HRESULT hr = AmsiScanBuffer(s_amsiContext, flatImageBytes, size, nullptr, nullptr, &result);
```

Comment:
What other HRESULTs can this return? Should we assume the payload is ok if `AmsiScanBuffer` returns `E_SOMETHING`?

Reply:
It only returns S_OK if it successfully scanned and anything else indicates it did not successfully scan (none of the AMSI providers succeeded). In that case, we treat it the same as if AMSI isn't available and do not block loading the bytes payload (matching full Framework behaviour).

Reply:
Sounds good - thanks for the clarification! :)